state/src/account.rs

```diff
@@ -0,0 +1,401 @@
+use cita_trie::db::DB;
+use cita_trie::trie::{PatriciaTrie, Trie};
+use ethereum_types::{H256, U256};
+use sha3::{Digest, Sha3_256};
+use std::collections::HashMap;
+
+pub const SHA3_EMPTY: H256 = H256([
+    167, 255, 198, 248, 191, 30, 215, 102, 81, 193, 71, 86, 160, 97, 214, 98, 245, 128, 255, 77,
+    228, 59, 73, 250, 130, 216, 10, 75, 128, 248, 67, 74,
+]);
+pub const SHA3_NULL_RLP: H256 = H256([
+    188, 32, 113, 164, 222, 132, 111, 40, 87, 2, 68, 127, 37, 137, 221, 22, 54, 120, 224, 151, 42,
+    138, 27, 13, 40, 176, 78, 213, 192, 148, 84, 127,
+]);
+
+#[derive(Debug)]
+pub struct Account {
+    balance: U256,
+    nonce: U256,
+    storage_root: H256,
+    code_hash: H256,
+}
+
+impl rlp::Encodable for Account {
+    fn rlp_append(&self, s: &mut rlp::RlpStream) {
+        s.begin_list(4)
+            .append(&self.nonce)
+            .append(&self.balance)
+            .append(&self.storage_root)
+            .append(&self.code_hash);
+    }
+}
+
+impl rlp::Decodable for Account {
+    fn decode(data: &rlp::Rlp) -> Result<Self, rlp::DecoderError> {
+        Ok(Account {
+            nonce: data.val_at(0)?,
+            balance: data.val_at(1)?,
+            storage_root: data.val_at(2)?,
+            code_hash: data.val_at(3)?,
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum CodeState {
+    Clean,
+    Dirty,
+}
+
+#[derive(Debug)]
+pub struct StateObject {
+    balance: U256,
+    nonce: U256,
+    storage_root: H256,
+    code_hash: H256,
+    code: Vec<u8>,
+    code_size: usize,
+    code_state: CodeState,
+    storage_changes: HashMap<H256, H256>,
+}
+
+impl From<Account> for StateObject {
+    fn from(account: Account) -> Self {
+        StateObject {
+            balance: account.balance,
+            nonce: account.nonce,
+            storage_root: account.storage_root,
+            code_hash: account.code_hash,
+            code: vec![],
+            code_size: 0,
+            code_state: CodeState::Clean,
+            storage_changes: HashMap::new(),
+        }
+    }
+}
+
+impl StateObject {
+    /// Create a new account.
+    /// NOTE: If contract account generated, make sure you use `init_code` on
+    /// this before `commit`ing.
+    pub fn new(balance: U256, nonce: U256) -> StateObject {
+        StateObject {
+            balance,
+            nonce,
+            storage_root: SHA3_NULL_RLP,
+            code_hash: SHA3_EMPTY,
+            code: vec![],
+            code_size: 0,
+            code_state: CodeState::Clean,
+            storage_changes: HashMap::new(),
+        }
+    }
+
+    /// Create a new account from rlp bytes.
+    /// Note: make sure you use `read_code` after this.
+    pub fn from_rlp(data: &[u8]) -> Result<StateObject, Error> {
+        let account: Account = rlp::decode(data)?;
+        Ok(account.into())
+    }
+
+    pub fn rlp(&self) -> Vec<u8> {
+        rlp::encode(&self.account())
+    }
+
+    pub fn account(&self) -> Account {
+        Account {
```

Comment:
recommend `pub fn as_account(&self) -> Account`

Reply:
I don't think this is necessary. StateObject is a wrap for account, so the account() is like a filed in StateObject, not `StateObject as Account`